src/plugins/gs-plugin-eos.c

```diff
@@ -220,12 +228,73 @@ gs_plugin_eos_refine_core_app (GsApp *app)
 	}
 }
 
+typedef struct _PopularBackgroundImageTileRequestData
+{
+	GsApp *app;
+	GsPlugin *plugin;
+	char *cache_filename;
+} PopularBackgroundImageTileRequestData;
+
+static void
+popular_background_image_tile_request_data_destroy (PopularBackgroundImageTileRequestData *data)
+{
+	g_free (data->cache_filename);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(PopularBackgroundImageTileRequestData,
+                              popular_background_image_tile_request_data_destroy);
+
+static void gs_plugin_eos_refine_popular_app(GsPlugin *plugin, GsApp *app);
+
+static void
+gs_plugin_eos_update_tile_image_from_filename (GsApp      *app,
+                                               const char *filename)
+{
+	g_autofree char *css = g_strdup_printf("background-image: url('%s')",
+	                                       filename);
+	gs_app_set_metadata (app, "GnomeSoftware::ImageTile-css", css);
+}
+
+static void
+gs_plugin_eos_tile_image_downloaded_cb (SoupSession *session,
+                                        SoupMessage *msg,
+                                        gpointer user_data)
+{
+	g_autoptr(PopularBackgroundImageTileRequestData) data = user_data;
+	g_autoptr(GError) error = NULL;
+
+	if (msg->status_code == SOUP_STATUS_CANCELLED)
+		return;
+
+	if (msg->status_code != SOUP_STATUS_OK) {
+		g_debug("Failed to download tile image corresponding to cache entry %s",
+		        data->cache_filename);
+		return;
+	}
+
+	/* Write out the cache image to disk */
+	if (!g_file_set_contents (data->cache_filename,
+	                          msg->response_body->data,
+	                          msg->response_body->length,
+	                          &error)) {
+		g_debug("Failed to write cache image %s, %s",
+		        data->cache_filename,
+		        error->message);
+		return;
+	}
+
+	gs_plugin_eos_refine_popular_app (data->plugin, data->app);
```

Comment:
> Also why do you call refine here, instead of setting the CSS like in the other case?

Thanks for catching that! I had meant to add that when I split out `gs_plugin_eos_update_image_from_filename` but obviously never got around to calling it here.